### PR TITLE
[AD-331] Adjust order of join operations

### PR DESCRIPTION
**Pull request by mitchell-elholm**
### Summary

Adjust order of join operations

### Description

Changed order that operations are added in the aggregation pipeline. $match and $addFields will now be placed before $unwind to improve performance.

### Additional Reviewers
@birschick-bq
@mitchell-elholm
@alexey-temnikov 
@andiem-bq

### CODE REVIEW

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java [hunk 1]
```diff
@@ -1030,28 +1037,27 @@ void testSameCollectionJoin() throws SQLException {
...
-                innerJoinResult.getAggregateOperations().get(1));
+                innerJoinResult.getAggregateOperations().get(0));
+        Assertions.assertEquals(
+                BsonDocument.parse(
+                        "{\"$unwind\": {\"path\": \"$array\", \"preserveNullAndEmptyArrays\": true, \"includeArrayIndex\": \"array_index_lvl_0\"}}"),
+                innerJoinResult.getAggregateOperations().get(2));
         Assertions.assertEquals(
                 BsonDocument.parse(
-                        "{\"$addFields\": {\"testCollection__id0\": {\"$cond\": [{\"$or\": [{\"$ifNull\": [\"$array.field\", false]}, {\"$ifNull\": [\"$array.field1\", false]}, {\"$ifNull\": [\"$array.field2\", false]}]}, \"$_id\", null]}, \"_id\": \"$_id\"}}"),
+                        "{ \"$unwind\": {"
```
Comment (andiemontoyeah) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java:1053]: assertion for operation 2 occurs twice while assertion for aggregation operation 1 is missing

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbJoin.java [hunk 2]
```diff
@@ -259,12 +236,37 @@ private void joinSameCollection(
...
+            default:
+                throw new IllegalArgumentException(
+                        SqlError.lookup(SqlError.UNSUPPORTED_JOIN_TYPE, getJoinType().name()));
+        }
         if (!renames.isEmpty()) {
             final String newFields = Util.toString(renames, "{", ", ", "}");
             final String aggregateString = "{ $addFields : " + newFields + "}";
             implementor.add(null, aggregateString);
         }
+        leftList.forEach(pair -> implementor.add(pair.left, pair.right));
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbJoin.java:267]: Nit: Probably needs a comment? `// Add operations from the left.`

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java [hunk 3]
```diff
@@ -1076,11 +1082,10 @@ void testSameCollectionJoin() throws SQLException {
                                 + "\"path\": \"$array\", "
                                 + "\"includeArrayIndex\" : \"array_index_lvl_0\", "
                                 + "\"preserveNullAndEmptyArrays\": true }}"),
-                leftJoinResult.getAggregateOperations().get(0));
+                leftJoinResult.getAggregateOperations().get(1));
```
Comment (birschick-bq) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java]: You want `.get(0)` here.

File: calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java [hunk 4]
```diff
@@ -1127,21 +1132,21 @@ void testComplexQueryWithSameCollectionJoin() throws SQLException {
         Assertions.assertEquals(14, result.getAggregateOperations().size());
         Assertions.assertEquals(
                 BsonDocument.parse(
+                        "{\"$addFields\": {\"testCollection__id0\": {\"$cond\": [{\"$or\": [{\"$ifNull\": [\"$array.field2\", false]}, " +
+                                "{\"$ifNull\": [\"$array.field\", false]}, {\"$ifNull\": [\"$array.field1\", false]}]}, \"$_id\", null]}, \"_id\": \"$_id\"}}"),
+                result.getAggregateOperations().get(1));
```
Comment (birschick-bq) [inline calcite-adapter/src/test/java/software/amazon/documentdb/jdbc/query/DocumentDbQueryMappingServiceTest.java:1140]: Missing test for `.get(0)` aggregate operation.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbJoin.java [hunk 5]
```diff
@@ -55,6 +55,7 @@
 import software.amazon.documentdb.jdbc.metadata.DocumentDbSchemaColumn;
 import software.amazon.documentdb.jdbc.metadata.DocumentDbSchemaTable;
 
+
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbJoin.java]: nit: unnecessary extra line.

**Review by birschick-bq — APPROVED** [review]

**Review by andiemontoyeah — APPROVED** [review]

**Pull request merged**